### DISCUSSION TITLE
run_edrumulus.sh fails at drum kit setup line 140

### DISCUSSION THREAD

**Discussion by AlienDiva** [corrados/edrumulus#94] [General]
Hello, I am setting up edrumulus software on a laptop with Ubuntu 22.04 server. The package requirements and compile process went without a drama but run_edrumulus.sh fails at drum kit setup line 140 with the following notable message:

We assume that you have created the edrumuluskit with edrumulus/tools/mixdown_kits.m or PearlMMX is present 

I've searched for PearlMMX kit and found nothing and have no idea how to "create the edrumuluskit with edrumulus/tools/mixdown_kits.m"

Running drumgizmo standalone with jack manually from the command line using DRSKit kit which I had already run_edrumulus.sh runs and I get the terminal gui.

I feel there's a whole bunch of something I'm missing with the kit setup so asking for help with clarification on the failed task. 

Thankyou kindly in advance.

**Comment by corrados** [comment]
The PearlMMX kit is based on samples I recorded from my acoustic kit. Unfortunately, these files are not (yet) available for download. Therefore, you cannot use it.

The way I am using this kit is to have a directory source_samples in the tools directory and then I call the Python script create_drumgizmo_kit.py to create the actual PearlMMX kit. The samples are in an "experimental" state since I did testing the positional sensing also on the toms. As soon as I am happy with the kit (and that will not be anytime soon), I plan to upload it to the official Drumgizmo page. I am already in contact with the main developer of Drumgizmo.

Anyway, the "old" way of creating a Drumgizmo kit for Edrumulus usage was to download the official Drumgizmo kits and run the mixdown_kits.m file in Octave. This script can either convert the entire kit (using, e.g., kit_select=3) or a selection of different instrument from different Drumgizmo kits (like it is configured right now). The intention of this script was to mix down the multi channel instruments of the official Drumgizmo kits to a stereo sample to reduce the size of the kits and also to have a simple way of playing these files without having to connect a DAW to mix all the channels.

I have to admit that I did not run that script mixdown_kits.m for a long time. I hope that it still works. If it works, it will create a subfolder drumuluskit in the tool directory which is then automatically used by the run_edrumulus.sh script.

**Reply by AlienDiva** [reply]
I can confirm that mixdown_kits.m octave script still works absolutely fine once I worked out what I was doing, the kit samples mix down to ./edrumuluskit/ if select_kit=N is set at the top of the script. I couldn't see how to select a kit as a command line argument so assuming that's not supported.

Without specifying a kit (kit_select=[]) the script borks and complains about file inclusion for the mapex_mars kit which I couldn't find.

So all good. Very useful tool, I've been fumbling around with sox in bash to convert multichannel samples to stereo for headless SBC.

**Reply by corrados** [reply]
The mapex_mars was also a recoding I have done with my old snare drum. It is actually not an entire kit but only the snare with positional sensing support. These recording are old and pretty bad so that I do not want to make these available at all.

**Reply by corrados** [reply]
I have now added the source samples for the PearlMMX kit. The zip file is added to the latest release 0.8: https://github.com/corrados/edrumulus/releases/tag/r0_8
Just unzip the files in edrumulus/tools/source_samples and run the create_drumgizmo_kit.py script.

**Reply by corrados** [reply]
I have also added the full PearlMMX drum kit now: https://github.com/corrados/edrumulus/releases/download/r0_8/PearlMMX.zip
It should (hopefully) work to unzip it on the tools directory and running the run_edrumulus.sh script.
I have created a Youtube video to showcase the sound of the PearlMMX kit: https://youtu.be/xHr5WekxPq0?si=nh9tXmqXq7vBZL9G

**Comment by AlienDiva** [comment]
Thankyou for your prompt and comprehensive answers as well as your excellent efforts into the project.

So hopefully I understand correctly, mixdown_kits.m is a MATLAB script to recreate kits suitable for rPi or other SBC and not a necessary step. Except for recreating 2 channel simplified version of a kit which does away with the requirements of interfacing with a DAW directly correct?

run_edrumulus.sh at the point of failure, which is loading of the nonexistent drumgizmo kits could in fact load a super basic kit (does not need to be perfect) which could be included with edumulus as you have suggested for the sake of continuity? Therein the user can explore options. 

The sample player also could be any other CLI based jack client, another option is sfizz_jack which is actively developed and while not perfect a reasonable simplified solution. I am looking into that further.

Anyhow, these are some of my ideas for today, I will say the MATLAB requirements for building kits is not my favourite but hopefully in time we can provide an alternative which is easily manageable.

Happy drumming and again may I express appreciation for your efforts towards the DIY community.

**Reply by corrados** [reply]
> mixdown_kits.m is a MATLAB script to recreate kits suitable for rPi or other SBC and not a necessary step. Except for recreating 2 channel simplified version of a kit which does away with the requirements of interfacing with a DAW directly correct?

That is correct.

> load a super basic kit (does not need to be perfect) which could be included with edumulus as you have suggested for the sake of continuity?

Good idea. I've added it to the [TODO list](https://github.com/corrados/edrumulus/blob/main/doc/TODO.md).

> another option is sfizz_jack which is actively developed

I have not tried that, yet. Do you know if it supports positional sensing?

> the MATLAB requirements for building kits is not my favourite but hopefully in time we can provide an alternative which is easily manageable.

What would you do differently? One option could be to do it in Python. I really like this language and have written some scripts already, although I'm a total beginner with Python :-).

> Happy drumming and again may I express appreciation for your efforts towards the DIY community.

Thank you :-).

**Reply by AlienDiva** [reply]
For positional sensing in sfizz, I believe it's doable using opcodes on_loccN and on_hiccN (not tested) See the example in the last post of the discussion you created here: 

https://linuxmusicians.com/viewtopic.php?t=22557

If I recall correctly, blending samples is also a possibility. I'll look it up again and get back.

Completely disregard my comment about matlab usage which I thought was just the kit generation, I recently discovered your testing scripts in tools/misc, so continue on. Fantastic work!

**Reply by AlienDiva** [reply]
I looked into positional sensing on sfizz further and assuming the midi is a single note plus control message value it can be done easily. Can you confirm that's the case or is there more to it?

If so it would be similar to  hihat openness articulations,.which definitely does work I have tested it. https://youtu.be/Chow4_5orXk?si=UivXJBWsTetZpY8G

Drumgizmo as far as I know does not support hihat control beyond open and closed but I see it's on their to-do.

**Reply by corrados** [reply]
> the midi is a single note plus control message value it can be done easily.

Yes, it is. A MIDI note plus an additional controller which gives the position.

> Drumgizmo as far as I know does not support hihat control beyond open and closed but I see it's on their to-do.

On my edrumulus branch in the Drumgizmo git repo, the hi-hat controller is working.

**Reply by AlienDiva** [reply]
Nice! Having hihat control was a the reason I was looking into sfizz over Drumgizmo.

So does edrumulus branch support snare position articulations and do any kits support both features?

**Reply by corrados** [reply]
Yes, it does snare position. But I have some experimental code in this branch which is not ready for merging to master. I have implemented some dynamic extender and do some non-linear modifications of the velocity to get more realistic dynamic. Also, for the hi-hat, I have implemented some experimental choke function for half-open hi-hat. This does work ok for some samples but might work really bad for other samples.

**Reply by corrados** [reply]
Here is a link to my Drumgizmo branch where you can see all my commits: http://cgit.drumgizmo.org/drumgizmo.git/log/?h=edrumulus

**Reply by corrados** [reply]
BTW, if you use my run_edrumulus.sh script, it will automatically checkout my branch, see: https://github.com/corrados/edrumulus/blob/26a325d1d87850e1e95e45824d7380b5733a2849/tools/run_edrumulus.sh#L116

**Reply by AlienDiva** [reply]
Do you have an example how to use the snare position articulations in Drumgizmo. I'm assuming something should be added to the instrument XML?

I did manage to get a basic test snare inner-to-outer hit with layer crossfade working in Sfizz.4 velocity layers and 2 articulations (centre and edge). The positional sensing continuous message on cc16 which it appears you are using, 127 for centre and 0 for outer, anything in between mixes the smaples. still needs work but it looks promising. Finding suitable kit samples is another story

**Reply by corrados** [reply]
The Drumgizmo main developer deva was so kind to create a snare instrument in Drumgizmo with full positional sensing support. You can download it here: https://drumgizmo.org/kits/snare-position.zip
If you open the file snare-position/snare-position.xml, you will find an example XML. The new parameter `position="0"` defines the position with a value between 0 to 1, where 0 is the center of the snare and 1 is the edge.

**Reply by AlienDiva** [reply]
Awesome thanks I'll check that out.

**Comment by 3hhh** [comment]
This stereo mix here should work too, shouldn't it?
https://drumgizmo.org/kits/CrocellKit/CrocellKit_Stereo_MIX.rar

Unfortunately I cannot get any sound from drumgizmo so far, but I fear it might not be related to the kit.
In fact I got sound once, but only after waiting for a very long time and only for certain Midi notes (40 worked, 38 didn't - even though the kit has mapped both). Subsequent tries failed again. Writing to a wav file with drumgizmo generates huge files that are way longer than the time which passed in reality. They're usually quiet as well...
I tried on two different machines with the same results so far.

**Reply by corrados** [reply]
> This stereo mix here should work too, shouldn't it?

Haven't tried but I guess it should work.
If you want to use it together with the run_edrumulus.sh, you have to adjust the Drumgizmo jack output names for this kit.

> Unfortunately I cannot get any sound from drumgizmo so far

If you have issues running Drumgizmo, I recommend to post a message in the following forum: [this forum](https://linuxmusicians.com/viewforum.php?f=55).

**Reply by 3hhh** [reply]
Got it running with jack, looks like pipewire & drumgizmo is a bad idea.

**Comment by corrados** [comment]
FYI, I have just made changes to the Edrumulus repo but also to the Drumgizmo edrumulus branch. If you use the [run_edrumulus.sh](https://github.com/corrados/edrumulus/blob/main/tools/run_edrumulus.sh) script, I recommend to delete the drumgizmo folder so that the script clones and makes it again to get the most recent Drumgizmo changes. Otherwise the [mixdown_kits.m](https://github.com/corrados/edrumulus/blob/main/tools/mixdown_kits.m) and [create_drumgizmo_kit.py](https://github.com/corrados/edrumulus/blob/main/tools/create_drumgizmo_kit.py) are no longer compatible to the old Drumgizmo version.

**Comment by 3hhh** [comment]
Thanks a lot, also for the Pearl MMX samples! :-)
